src/basicfuns.jl

```diff
@@ -31,7 +41,19 @@ The [logistic](https://en.wikipedia.org/wiki/Logistic_function) sigmoid function
 
 Its inverse is the [`logit`](@ref) function.
 """
-logistic(x::Real) = inv(exp(-x) + one(x))
+function logistic(x::Real)
+    e = exp(x)
+    lower, upper = _logistic_bounds(x)
+    ifelse(
```

Comment:
Seems like a quite convoluted way of writing it over just `if`, `else` etc?

```jl
function logistic(x::Real)
    e = exp(x)
    lower, upper = _logistic_bounds(x)
    if x < lower
        return zero(x)
    elseif x > upper
        return one(x)
    else
        return e / (one(x) + e)
    end
end
```

Reply:
Probably worth checking `@code_llvm`; it's true that LLVM now does the conversion of `if-else` => `ifelse` automatically pretty well these days.

Reply:
They seem to produce different output at the `@code_llvm` and `@code_native` stages consistent with my earlier experience that `ifelse` generates [LLVM `select`](https://releases.llvm.org/6.0.0/docs/LangRef.html#select-instruction) and `if` generates [LLVM `br`](https://releases.llvm.org/6.0.0/docs/LangRef.html#br-instruction), but benchmarking suggests their performance is very similar in scalar applications and broadcasting: https://gist.github.com/johnmyleswhite/548f4eb18a028a237d52ae06811ca33c

Note that the benchmarks also suggest that changing from the old implementation to the new implementation causes performance to drop because of processing subnormals in a mirror image: the old formulation was fast for `z = -710.0` and slow for `z = +710.0`, whereas the new formulation is slow for `z = -710.0` and fast for `z = +710.0`.

Happy to choose whichever form people prefer. I tend to defer to the side of assuming `select` is more likely to play well with SIMD, but it doesn't seem like that issue applies in this case for the benchmarks I've run.

Reply:
> Happy to choose whichever form people prefer.

Personally, I don't really care, just felt like the `if` version was easier to read and I couldn't find a benchmark where it was slower. I guess if you correctly predict the branch version is faster and otherwise the branchless is faster? :man_shrugging:

Reply:
Ok, I think in the absence of useful evidence either way I'll maintain my existing superstitious faith in the importance of the LLVM `select` instruction. Revising in the future would be trivial.